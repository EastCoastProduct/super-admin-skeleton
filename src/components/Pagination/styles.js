import { StyleSheet } from 'aphrodite/no-important';

export default StyleSheet.create({
  item: {
    display: 'inline-block',
    padding: 5,
    verticalAlign: 'middle',
  },
  selected: {
    fontWeight: 700,
  },
  pagination: {
    marginTop: 10,
  },
});
