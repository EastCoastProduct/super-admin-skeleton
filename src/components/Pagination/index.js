import React, { Component, PropTypes } from 'react';
import { css } from 'aphrodite/no-important';
import Button from '../Button';
import styles from './index.style';

export default class Pagination extends Component {
  static propTypes = {
    currentPage: PropTypes.number.isRequired,
    itemsPerPage: PropTypes.number.isRequired,
    onPaginationChange: PropTypes.func.isRequired,
    style: PropTypes.object,
    total: PropTypes.number.isRequired,
  };

  componentWillMount() {
    this.calculatePagination();
  }

  componentWillReceiveProps(nextProps) {
    this.calculatePagination(nextProps);
  }

  calculatePagination(nextProps) {
    const { itemsPerPage, total } = nextProps || this.props;
    this.numOfPages = Math.ceil(total / itemsPerPage);
  }

  handleChangePage(page) {
    const { onPaginationChange } = this.props;
    onPaginationChange(page);
  }

  clickHandler(i) {
    return () => this.handleChangePage(i);
  }

  renderPages() {
    if (this.numOfPages < 2) return false;
    const { currentPage } = this.props;
    const elems = [];
    let i;
    let j;

    // calculate start/end of pagination numbers if pagination has over 5 pages
    if (currentPage - 2 < 1) {
      i = 1;
      j = 5;
    } else if (currentPage + 2 > this.numOfPages) {
      i = this.numOfPages - 4;
      j = this.numOfPages;
    } else {
      i = currentPage - 2;
      j = currentPage + 2;
    }

    // calculate if pagination has 5 or less pages, reset previous calculations
    if (this.numOfPages <= 5) {
      i = 1;
      j = this.numOfPages;
    }

    for (i; i <= j; i += 1) {
      elems.push(
        <li
          className={css(styles.item, i === currentPage && styles.selected)}
          key={i}
        >
          {i === currentPage ? i :
            <Button
              empty
              type="button"
              onClick={this.clickHandler(i)}
            >{i}</Button>
          }
        </li>
      );
    }

    return elems;
  }

  renderPaginationInfo() {
    const { currentPage, itemsPerPage, total } = this.props;

    if (total === 1) return 'Showing 1 result';
    if (this.numOfPages <= 1) return `Showing all of ${total} results`;

    const currentStart = 1 + ((currentPage - 1) * itemsPerPage);
    let currentEnd = itemsPerPage + ((currentPage - 1) * itemsPerPage);
    currentEnd = currentEnd > total ? total : currentEnd;
    return `Showing ${currentStart}-${currentEnd} of ${total} results`;
  }

  render() {
    const { currentPage, style, total } = this.props;

    return (
      total > 0 &&
        <div className={css(style && style)}>
          <p>{this.renderPaginationInfo()}</p>
          {this.numOfPages > 1 &&
            <ul className={css(styles.pagination)}>
              { currentPage > 3 && this.numOfPages > 5 &&
                <li className={css(styles.item)}>
                  <Button
                    empty
                    type="button"
                    onClick={() => this.handleChangePage(1)}
                  >
                    <i className="fa fa-angle-double-left" />
                  </Button>
                </li>
              }
              { currentPage !== 1 &&
                <li className={css(styles.item)}>
                  <Button
                    empty
                    type="button"
                    onClick={() => this.handleChangePage(currentPage - 1)}
                  >
                    <i className="fa fa-angle-left" />
                  </Button>
                </li>
              }
              {this.renderPages()}
              {currentPage !== this.numOfPages &&
                <li className={css(styles.item)}>
                  <Button
                    empty
                    type="button"
                    onClick={() => this.handleChangePage(currentPage + 1)}
                  >
                    <i className="fa fa-angle-right" />
                  </Button>
                </li>
              }
              {currentPage < this.numOfPages - 2 && this.numOfPages > 5 &&
                <li className={css(styles.item)}>
                  <Button
                    empty
                    type="button"
                    onClick={() => this.handleChangePage(this.numOfPages)}
                  >
                    <i className="fa fa-angle-double-right" />
                  </Button>
                </li>
              }
            </ul>
          }
        </div>
    );
  }
}
