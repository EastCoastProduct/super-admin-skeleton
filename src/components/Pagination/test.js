import React from 'react';
import { shallow } from 'enzyme';
import { StyleSheetTestUtils } from 'aphrodite/no-important';
import Pagination from './';

describe('Pagination component', () => {
  beforeEach(() => {
    StyleSheetTestUtils.suppressStyleInjection();
  });
  afterEach(() => {
    StyleSheetTestUtils.clearBufferAndResumeStyleInjection();
  });

  const mockPaginationChange = jest.fn();
  const wrapper = shallow(
    <Pagination
      currentPage={1}
      itemsPerPage={10}
      onPaginationChange={mockPaginationChange}
      total={127}
    />
  );
  const instance = wrapper.instance();

  it('calculatePagination method', () => {
    instance.calculatePagination({
      itemsPerPage: 5,
      total: 44,
    });
    expect(instance.numOfPages).toBe(9);

    instance.calculatePagination();
    expect(instance.numOfPages).toBe(13);
  });

  it('handleChangePage method', () => {
    instance.handleChangePage(5);

    expect(mockPaginationChange).toHaveBeenCalledWith(5);
  });

  it('clickHandler method', () => {
    instance.handleChangePage = jest.fn();
    const handlerFunction = instance.clickHandler(3);
    handlerFunction();

    expect(instance.handleChangePage).toHaveBeenCalledWith(3);
  });

  it('renderPages method', () => {
    instance.numOfPages = 1;
    expect(instance.renderPages()).toBeFalsy();

    instance.numOfPages = 3;
    expect(instance.renderPages().length).toBe(3);

    instance.numOfPages = 7;
    let elems = instance.renderPages();
    expect(elems.length).toBe(5);
    expect(elems[0].key).toBe('1');
    expect(elems[4].key).toBe('5');

    wrapper.setProps({
      currentPage: 5,
    });
    instance.numOfPages = 13;
    elems = instance.renderPages();
    expect(elems.length).toBe(5);
    expect(elems[0].key).toBe('3');
    expect(elems[4].key).toBe('7');
  });

  it('renderPaginationInfo method', () => {
    expect(instance.renderPaginationInfo())
      .toBe('Showing 41-50 of 127 results');

    wrapper.setProps({
      currentPage: 1,
      total: 1,
    });
    expect(instance.renderPaginationInfo()).toBe('Showing 1 result');

    wrapper.setProps({
      total: 5,
    });
    expect(instance.renderPaginationInfo()).toBe('Showing all of 5 results');
  });

  it('should call methods inside render function', () => {
    instance.handleChangePage = jest.fn();
    wrapper.setProps({
      currentPage: 5,
      total: 127,
    });
    wrapper.find('.fa-angle-double-left').parent().simulate('click');
    expect(instance.handleChangePage).toHaveBeenCalledWith(1);

    wrapper.find('.fa-angle-left').parent().simulate('click');
    expect(instance.handleChangePage).toHaveBeenCalledWith(4);

    wrapper.find('.fa-angle-right').parent().simulate('click');
    expect(instance.handleChangePage).toHaveBeenCalledWith(6);

    wrapper.find('.fa-angle-double-right').parent().simulate('click');
    expect(instance.handleChangePage).toHaveBeenCalledWith(13);
    expect(instance.handleChangePage).toHaveBeenCalledTimes(4);
  });
});
