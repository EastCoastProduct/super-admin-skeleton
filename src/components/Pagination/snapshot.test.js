import React from 'react';
import renderer from 'react-test-renderer';
import { StyleSheet, StyleSheetTestUtils } from 'aphrodite/no-important';
import Pagination from './';

describe('Pagination component snapshot', () => {
  beforeEach(() => {
    StyleSheetTestUtils.suppressStyleInjection();
  });
  afterEach(() => {
    StyleSheetTestUtils.clearBufferAndResumeStyleInjection();
  });

  it('renders required data with current page 1', () => {
    const tree = renderer.create(
      <Pagination
        currentPage={1}
        itemsPerPage={10}
        onPaginationChange={jest.fn()}
        total={127}
      />
    );

    expect(tree.toJSON()).toMatchSnapshot();
  });

  it('renders required data with current page 5', () => {
    const tree = renderer.create(
      <Pagination
        currentPage={5}
        itemsPerPage={10}
        onPaginationChange={jest.fn()}
        total={93}
      />
    );

    expect(tree.toJSON()).toMatchSnapshot();
  });

  it('renders required data with current page being last', () => {
    const tree = renderer.create(
      <Pagination
        currentPage={8}
        itemsPerPage={10}
        onPaginationChange={jest.fn()}
        total={78}
      />
    );

    expect(tree.toJSON()).toMatchSnapshot();
  });

  it('renders required data with less than 5 pages', () => {
    const tree = renderer.create(
      <Pagination
        currentPage={2}
        itemsPerPage={10}
        onPaginationChange={jest.fn()}
        total={30}
      />
    );

    expect(tree.toJSON()).toMatchSnapshot();
  });

  it('renders just message with extra style for small number of items', () => {
    const style = StyleSheet.create({
      pagination: {
        margin: 10,
        padding: 10,
      },
    });
    const tree = renderer.create(
      <Pagination
        currentPage={1}
        itemsPerPage={10}
        onPaginationChange={jest.fn()}
        style={style.pagination}
        total={5}
      />
    );

    expect(tree.toJSON()).toMatchSnapshot();
  });

  it('renders just message for 1 item', () => {
    const tree = renderer.create(
      <Pagination
        currentPage={1}
        itemsPerPage={10}
        onPaginationChange={jest.fn()}
        total={1}
      />
    );

    expect(tree.toJSON()).toMatchSnapshot();
  });

  it('doesn\'t render anything for 0 items', () => {
    const tree = renderer.create(
      <Pagination
        currentPage={1}
        itemsPerPage={10}
        onPaginationChange={jest.fn()}
        total={0}
      />
    );

    expect(tree.toJSON()).toMatchSnapshot();
  });
});
