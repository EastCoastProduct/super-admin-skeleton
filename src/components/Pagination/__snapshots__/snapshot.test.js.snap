exports[`Pagination component snapshot doesn't render anything for 0 items 1`] = `null`;

exports[`Pagination component snapshot renders just message for 1 item 1`] = `
<div
  className="">
  <p>
    Showing 1 result
  </p>
</div>
`;

exports[`Pagination component snapshot renders just message with extra style for small number of items 1`] = `
<div
  className="">
  <p>
    Showing all of 5 results
  </p>
</div>
`;

exports[`Pagination component snapshot renders required data with current page 1 1`] = `
<div
  className="">
  <p>
    Showing 1-10 of 127 results
  </p>
  <ul
    className="pagination_1ka7v26">
    <li
      className="item_ifw64p-o_O-selected_1mzzwq6">
      1
    </li>
    <li
      className="item_ifw64p">
      <button
        className="empty_1b6lvt9"
        onClick={[Function]}
        type="button">
        2
      </button>
    </li>
    <li
      className="item_ifw64p">
      <button
        className="empty_1b6lvt9"
        onClick={[Function]}
        type="button">
        3
      </button>
    </li>
    <li
      className="item_ifw64p">
      <button
        className="empty_1b6lvt9"
        onClick={[Function]}
        type="button">
        4
      </button>
    </li>
    <li
      className="item_ifw64p">
      <button
        className="empty_1b6lvt9"
        onClick={[Function]}
        type="button">
        5
      </button>
    </li>
    <li
      className="item_ifw64p">
      <button
        className="empty_1b6lvt9"
        onClick={[Function]}
        type="button">
        <i
          className="fa fa-angle-right" />
      </button>
    </li>
    <li
      className="item_ifw64p">
      <button
        className="empty_1b6lvt9"
        onClick={[Function]}
        type="button">
        <i
          className="fa fa-angle-double-right" />
      </button>
    </li>
  </ul>
</div>
`;

exports[`Pagination component snapshot renders required data with current page 5 1`] = `
<div
  className="">
  <p>
    Showing 41-50 of 93 results
  </p>
  <ul
    className="pagination_1ka7v26">
    <li
      className="item_ifw64p">
      <button
        className="empty_1b6lvt9"
        onClick={[Function]}
        type="button">
        <i
          className="fa fa-angle-double-left" />
      </button>
    </li>
    <li
      className="item_ifw64p">
      <button
        className="empty_1b6lvt9"
        onClick={[Function]}
        type="button">
        <i
          className="fa fa-angle-left" />
      </button>
    </li>
    <li
      className="item_ifw64p">
      <button
        className="empty_1b6lvt9"
        onClick={[Function]}
        type="button">
        3
      </button>
    </li>
    <li
      className="item_ifw64p">
      <button
        className="empty_1b6lvt9"
        onClick={[Function]}
        type="button">
        4
      </button>
    </li>
    <li
      className="item_ifw64p-o_O-selected_1mzzwq6">
      5
    </li>
    <li
      className="item_ifw64p">
      <button
        className="empty_1b6lvt9"
        onClick={[Function]}
        type="button">
        6
      </button>
    </li>
    <li
      className="item_ifw64p">
      <button
        className="empty_1b6lvt9"
        onClick={[Function]}
        type="button">
        7
      </button>
    </li>
    <li
      className="item_ifw64p">
      <button
        className="empty_1b6lvt9"
        onClick={[Function]}
        type="button">
        <i
          className="fa fa-angle-right" />
      </button>
    </li>
    <li
      className="item_ifw64p">
      <button
        className="empty_1b6lvt9"
        onClick={[Function]}
        type="button">
        <i
          className="fa fa-angle-double-right" />
      </button>
    </li>
  </ul>
</div>
`;

exports[`Pagination component snapshot renders required data with current page being last 1`] = `
<div
  className="">
  <p>
    Showing 71-78 of 78 results
  </p>
  <ul
    className="pagination_1ka7v26">
    <li
      className="item_ifw64p">
      <button
        className="empty_1b6lvt9"
        onClick={[Function]}
        type="button">
        <i
          className="fa fa-angle-double-left" />
      </button>
    </li>
    <li
      className="item_ifw64p">
      <button
        className="empty_1b6lvt9"
        onClick={[Function]}
        type="button">
        <i
          className="fa fa-angle-left" />
      </button>
    </li>
    <li
      className="item_ifw64p">
      <button
        className="empty_1b6lvt9"
        onClick={[Function]}
        type="button">
        4
      </button>
    </li>
    <li
      className="item_ifw64p">
      <button
        className="empty_1b6lvt9"
        onClick={[Function]}
        type="button">
        5
      </button>
    </li>
    <li
      className="item_ifw64p">
      <button
        className="empty_1b6lvt9"
        onClick={[Function]}
        type="button">
        6
      </button>
    </li>
    <li
      className="item_ifw64p">
      <button
        className="empty_1b6lvt9"
        onClick={[Function]}
        type="button">
        7
      </button>
    </li>
    <li
      className="item_ifw64p-o_O-selected_1mzzwq6">
      8
    </li>
  </ul>
</div>
`;

exports[`Pagination component snapshot renders required data with less than 5 pages 1`] = `
<div
  className="">
  <p>
    Showing 11-20 of 30 results
  </p>
  <ul
    className="pagination_1ka7v26">
    <li
      className="item_ifw64p">
      <button
        className="empty_1b6lvt9"
        onClick={[Function]}
        type="button">
        <i
          className="fa fa-angle-left" />
      </button>
    </li>
    <li
      className="item_ifw64p">
      <button
        className="empty_1b6lvt9"
        onClick={[Function]}
        type="button">
        1
      </button>
    </li>
    <li
      className="item_ifw64p-o_O-selected_1mzzwq6">
      2
    </li>
    <li
      className="item_ifw64p">
      <button
        className="empty_1b6lvt9"
        onClick={[Function]}
        type="button">
        3
      </button>
    </li>
    <li
      className="item_ifw64p">
      <button
        className="empty_1b6lvt9"
        onClick={[Function]}
        type="button">
        <i
          className="fa fa-angle-right" />
      </button>
    </li>
  </ul>
</div>
`;
